### PR TITLE
ARROW-8993: [Rust] support reading non-seekable sources

### PR DESCRIPTION
**Pull request by zeapo**
A proposal of an implementation to read `json.gz` files. The idea is to along `.json` files, support gziped ones. 

The implementation of `json::ReadBuilder` heavily rely on `File`. My first approach would have been to make another version of the builder that would rely on a `BufReader<Box<dyn Read>>` and have the original one delegate to this the final build.

I went with another approach, in this case I've added an option to specify if the file is gziped or not. In this case, the original implementation works as defined, however requires a switch of type as per https://github.com/apache/arrow/compare/master...zeapo:feature/json-gzip?expand=1#diff-3fa6fc3c0ee201f5a3a1a5d25d0062ffR966-R967 

I've got no religion on this. Suggestions are welcome, to be honest, I would love that the `infer_json_schema` be public , which would make all of this code unnecessary as it would delegate the builder process to the users who care about these features.

3 options :)

### CODE REVIEW

**Comment by zeapo** [comment]
After sleeping on it, here's the 3rd option (exposing infer schema with BufReader): https://github.com/apache/arrow/compare/master...zeapo:feature/buffer-exposed

The advantage is that it does not need the dependency on gzip and has little impact on the api. It also opens the way to have other formats support.

**Comment by nevi-me** [comment]
> ...I would love that the `infer_json_schema` be public , which would make all of this code unnecessary as it would delegate the builder process to the users who care about these features.

Hi @zeapo, firstly, thanks for working on this. If exposing the function as public is helpful, please go ahead.

**Comment by nevi-me** [comment]
> It also opens the way to have other formats support

Do you mean other compression formats?

> However, this needs for the user to do manual seek (if the BufRead is on a File) and does not rely on the Builder. Unless a new builder with buffer is added.

I suppose the broader question to ask would be whether we'd like to support reading (and writing) compressed data. There have also been some changes on the `arrow::csv` side, such as allowing inference of multiple files, which might also be convenient to have in `arrow::json`.

What's your opinion on the 2 above? We need not implement them as part of this PR, but I suppose some user feedback is helpful :)

I'm still pro returning the reader back to the start, or is there a performance impact in doing so? I wouldn't want to place the burden of seeking on the user, because I'd expect the common inference case to be getting the schema then reading the file.

```rust
    // return the reader seek back to the start	
    reader.into_inner().seek(SeekFrom::Start(0))?;
```

**Comment by zeapo** [comment]
Thanks for your feedback.

> Do you mean other compression formats?

Yes, but not only. This would allow also to have different filesystems (like S3) where using `File` is not possible, and would make much more sense to have `dyn Read` (in rusoto they use ByteStream that implements Read and AsyncRead)

> There have also been some changes on the arrow::csv side, such as allowing inference of multiple files, which might also be convenient to have in arrow::json

That would be great. This would help when there are multiple small file that are written without a fixed batch size (small & big files) and data would be scattered across multiple files. If you have a JIRA issue for this I can take a look at it :)

> I'm still pro returning the reader back to the start, or is there a performance impact in doing so? I wouldn't want to place the burden of seeking on the user, because I'd expect the common inference case to be getting the schema then reading the file.

I agree that placing the burden on the user is a bad idea. However, there are situations where we just can't seek back to start (s3 is one example). Maybe a specific implementation for `Seek + Read`, that would do the seek back to start, and one for `Read` only, that would not. However... this would need the use of specialization, so more nightly dependencies.

Not really sure :/

**Comment by zeapo** [comment]
I've updated the little poc: https://github.com/apache/arrow/compare/master...zeapo:feature/buffer-exposed

The main changes are:
```rust
pub fn infer_json_schema_from_seekable<R: Read + Seek>(
   source: &mut R,
   max_read_records: Option<usize>
) -> Result<Arc<Schema>>
```
and
```rust
pub fn infer_json_schema_from_reader<R: Read>(
   reader: &mut BufReader<R>,
   max_read_records: Option<usize>
) -> Result<Arc<Schema>>
```
And the fact the `build` requires a `Seek` too. This makes the builder `File` agnostic, and relies only on Seek.

Now, if the user (i.e. me? ^_^) wants to use a non-seekable, I'd use the infer on the buffer, handle the memory myself and pass the schema I inferred.

**Comment by nevi-me** [comment]
> I agree that placing the burden on the user is a bad idea. However, there are situations where we just can't seek back to start (s3 is one example). Maybe a specific implementation for `Seek + Read`, that would do the seek back to start, and one for `Read` only, that would not. However... this would need the use of specialization, so more nightly dependencies.

Okay, in that case I could support not seeking back to the start of the input. One downside though is that in the case where no schema is supplied, and the readers (csv and json) infer the schema, we do need to reset the input to its starting position. I've briefly looked at the csv code, so if it's doable, we could find a solution.

Regarding specialization, we are already dependent on it, with low likelihood of this changing soon; so I'd say it's an option.

With regards to `dyn Read`, please have a look at the `arrow::csv::reader` code. We already support a `BufReader<R: Read>`, so I think we can implement the same without boxing the reader in the way that you've done so far. The only/primary reason why we still use `File` in `arrow::json` is that nobody's needed to use something else, or at least raised the issue.
___

One dramatic alternative would be to always require a schema, and leave inference to the user. We could then consume the buffer reader (`reader: mut BufReader<R>` instead of `reader: &mut BufReader<R>`), so that we don't leave the user with a file handle that's already partially/fully consumed.

**Comment by zeapo** [comment]
> One dramatic alternative would be to always require a schema, and leave inference to the user. We could then consume the buffer reader (reader: mut BufReader<R> instead of reader: &mut BufReader<R>), so that we don't leave the user with a file handle that's already partially/fully consumed.

I like this :+1:

**Comment by houqp** [comment]
> One dramatic alternative would be to always require a schema, and leave inference to the user. We could then consume the buffer reader (reader: mut BufReader<R> instead of reader: &mut BufReader<R>), so that we don't leave the user with a file handle that's already partially/fully consumed.

this is the approach csv reader took as well, so another benefit from doing this would be consistent user experience.

**Comment by zeapo** [comment]
Would it make sense to implement for json reader the same interfaces that are on csv reader, i.e. `Read + Seek` for `infer_file_schema`, and a `json::Reader::from_buf_reader()`.

Then in another PR add an interface on `Read` only? In both json and csv. That would mean creating a new `infer_reader_schema` that takes a `Read` (rather than `Read + Seek`).

**Comment by houqp** [comment]
Looks like we have to remove Seek constraints in order to support compressed format across the board. If this is the plan here as well, then I recommend just implement Read for json in this PR (with extra Seek helper function if you prefer). Then send a follow up PR to update csv reader.

**Comment by zeapo** [comment]
As per @houqp suggestion, here are the changes:

-  match csv's infer schema:
    - CSV: https://github.com/apache/arrow/blob/master/rust/arrow/src/csv/reader.rs#L90-L95 
    - JSON: https://github.com/apache/arrow/pull/7309/files#diff-3fa6fc3c0ee201f5a3a1a5d25d0062ffR174-R177
    - Mismatch: notice that the json's one return an `Arc<Schema>`. I don't know if we should make the csv reader match this or the other way arround. Especially that `Reader::new` takes `Arc<Schema>`
    - new function for seekable: https://github.com/apache/arrow/pull/7309/files?diff=split&w=1#diff-3fa6fc3c0ee201f5a3a1a5d25d0062ffR159-R168
- match csv's `Reader::from_buf_reader`:
    - csv: https://github.com/apache/arrow/blob/master/rust/arrow/src/csv/reader.rs#L235-L242
    - json: https://github.com/apache/arrow/pull/7309/files#diff-3fa6fc3c0ee201f5a3a1a5d25d0062ffR377-R382
- match csv's `Reader::new`:
    - csv: https://github.com/apache/arrow/blob/master/rust/arrow/src/csv/reader.rs#L198-L214
    - json: https://github.com/apache/arrow/pull/7309/files#diff-3fa6fc3c0ee201f5a3a1a5d25d0062ffR341-R347

**Review by houqp — APPROVED** [review]

**Comment by nevi-me** [comment]
> As per @houqp suggestion, here are the changes:

Thanks, I'll start reviewing them

File: rust/arrow/src/json/reader.rs [hunk 1]
```diff
@@ -157,9 +156,26 @@ fn generate_schema(spec: HashMap<String, HashSet<DataType>>) -> Result<Arc<Schem
 /// `max_read_records` controlling the maximum number of records to read.
 ///
 /// If `max_read_records` is not set, the whole file is read to infer its field types.
-fn infer_json_schema(file: File, max_read_records: Option<usize>) -> Result<Arc<Schema>> {
+pub fn infer_json_schema_from_seekable<R: Read + Seek>(
```
Comment (nevi-me) [inline rust/arrow/src/json/reader.rs:159]: May you please add a note on why a user would want/need to use this vs the non-seekable one

File: rust/arrow/src/json/reader.rs [hunk 2]
```diff
@@ -364,6 +370,24 @@ impl<R: Read> Reader<R> {
         }
     }
 
+    /// Create a new JSON Reader from a `BufReader<R: Read>`
+    ///
+    /// If reading a `File`, you can customise the Reader, such as to enable schema
```
Comment (nevi-me) [inline rust/arrow/src/json/reader.rs:375]: This doesn't flow well. Perhaps:

```
To customi{s|z}e the schema, such as to enable schema inference, use `ReaderBuilder`
```

File: rust/arrow/src/json/reader.rs [hunk 3]
```diff
@@ -1037,62 +1065,73 @@ mod tests {
...
-        let dd = dd.values();
-        let dd = dd.as_any().downcast_ref::<StringArray>().unwrap();
-        assert_eq!(7, dd.len());
-        assert_eq!(false, dd.is_valid(1));
-        assert_eq!("text", dd.value(2));
-        assert_eq!("1", dd.value(3));
-        assert_eq!("false", dd.value(4));
-        assert_eq!("array", dd.value(5));
-        assert_eq!("2.4", dd.value(6));
+        let mut file = File::open("test/data/mixed_arrays.json.gz").unwrap();
```
Comment (nevi-me) [inline rust/arrow/src/json/reader.rs:1068]: Will this mean that we're then not supporting zipped files for now? I'm fine with that.

**Review by nevi-me — CHANGES_REQUESTED** [review]
Minor documentation changes, but otherwise happy

File: rust/arrow/src/json/reader.rs (hunk 3 above)
Comment (houqp) [reply rust/arrow/src/json/reader.rs:1068]: yeah, i feel like compression support should be handled by higher level interface that takes filenames/paths as inputs.

Comment (zeapo) [reply rust/arrow/src/json/reader.rs:1068]: That would be great. I've seen that the C++ and Python libs have a filesystem interface. Maybe we could add that first, then work from there into adding mime-type and compression support?

File: rust/arrow/src/json/reader.rs (hunk 2 above)
Comment (zeapo) [reply rust/arrow/src/json/reader.rs]: changed :+1:

File: rust/arrow/src/json/reader.rs (hunk 1 above)
Comment (zeapo) [reply rust/arrow/src/json/reader.rs:175]: Updated with additional examples and other tests at the end. could you please check

**Review by nevi-me — APPROVED** [review]
Thanks, happy with the changes

**Comment by nevi-me** [comment]
Hi @zeapo, can you please rebase to fix the merge conflict? I'll merge this PR after that

**Comment by zeapo** [comment]
I'll fix the build tonight.

**Comment by zeapo** [comment]
Hey @nevi-me, it should be done ^_^

**Pull request closed without merging**